Chocolatey Addin for Cake

gep13 opened the pull request:
Choco commands that will be included in this first release of Chocolatey support

- [x] Install
- [x] Pack
- [x] Pin
- [x] Push
- [x] ApiKey/SetApiKey
- [x] Source/Sources
- [x] Update/Upgrade
- [x] Feature/Features
- [x] Config
- [x] Documentation

*NOTE:* Only the commands that apply can action will be included in this release, as these are the most likely commands that are needed within a build script.  For example, only `choco config set` will be implemented, where as `choco config get` will not.

Choco commands which are unlikely to be implemented in this PR:

* Search/List
* Outdated
* Unpackself
* New
* Uninstall
* Version

gep13 commented:
@devlead @patriksvensson I will fix up the CA error in the build

devlead commented on src/Cake.Common/Tools/Chocolatey/ChocolateyToolResolver.cs:
```diff
@@ -0,0 +1,80 @@
+    {
+        private readonly IFileSystem _fileSystem;
+        private readonly ICakeEnvironment _environment;
+        private FilePath _chocolateyPath;
```
This isn't used anywhere.

gep13 replied:
Yip, that is the CA error that I mentioned above, will fix that up.

For this part, I was comparing between SignToolResolver, and NugetToolResolver, and both do things in a slightly different way, and I ended up with an amalgamation of them both :smile:

devlead replied:
No worries, looks pretty solid otherwise, I will take a look again when issues fixed.

devlead commented on src/Cake.Common/Tools/Chocolatey/Pack/ChocolateyPacker.cs:
```diff
@@ -0,0 +1,158 @@
+            // Nuspec file
+            builder.AppendQuoted(nuspecFilePath.MakeAbsolute(_environment).FullPath);
+
+
```
Only one line break

gep13 commented:
@devlead @patriksvensson can I get some thoughts on the recent commit added to this PR?

I have changed it to actually use the command line properties for choco, which are somewhat different from nuget.  I had everything working, but then I removed the BasePath property from the ChocolateyPackSettings as this is not a command line argument for choco, and now when I hit the:

```
_roslynSession.Execute(code);
```

in the RoslynScriptSession.cs, I get an exception:

```
c:/temp/build.cake(33,9): error CS0117: 'Cake.Common.Tools.Chocolatey.Pack.ChocolateyPackSettings' does not contain a definition for 'BasePath'
```

But I don't see how/where that we be coming from.  Any ideas?

patriksvensson commented:
@gep13 Sure, what does the `c:/temp/build.cake` file look like?

gep13 commented:
@patriksvensson doh! double doh! triple doh!

As soon as you said that, I knew what the problem was...

Since I originally had those properties, they were in my cake file, so I am guessing Roslyn was trying to map those properties to the ChocolateyPackSettings, and failing.  This is the file I was/am using:

```
//////////////////////////////////////////////////////////////////////
// ARGUMENTS
//////////////////////////////////////////////////////////////////////

var target = Argument("target", "Default");
var configuration = Argument("configuration", "Release");

//////////////////////////////////////////////////////////////////////
// PREPARATION
//////////////////////////////////////////////////////////////////////

// Get whether or not this is a local build.
var local = BuildSystem.IsLocalBuild;
var isRunningOnWindows = IsRunningOnWindows();

// Get version.
var version = "1.2.3.4";

// Define directories.
var buildDir = Directory("C:\\temp\\CakeTemp");

//////////////////////////////////////////////////////////////////////
// TASKS
//////////////////////////////////////////////////////////////////////

Task("Create-Chocolatey-Packages")
    .WithCriteria(() => isRunningOnWindows)
    .Does(() =>
{
    // Create Chocolatey package.
    ChocolateyPack("C:\\temp\\CakeTemp\\caketemp.nuspec", new ChocolateyPackSettings {
        Version = version,
        BasePath = buildDir,
        OutputDirectory = buildDir
    });
});

//////////////////////////////////////////////////////////////////////
// TASK TARGETS
//////////////////////////////////////////////////////////////////////

Task("Package")
    .IsDependentOn("Create-Chocolatey-Packages");

Task("Default")
    .IsDependentOn("Package");

//////////////////////////////////////////////////////////////////////
// EXECUTION
//////////////////////////////////////////////////////////////////////

RunTarget(target);
```

Basically a butchered version of the build.cake file used to build cake.

gep13 commented:
Removing those properties now makes it work :+1:

patriksvensson commented:
Ah, perfect :+1:

devlead commented:
@gep13 Got some styling issues with the documentation tags.

gep13 commented:
> @gep13 Got some styling issues with the documentation tags.

@devlead doh! Fixed.

devlead commented on src/Cake.Common/Tools/Chocolatey/Pack/ChocolateyPacker.cs:
```diff
@@ -0,0 +1,185 @@
+            if (settings.ExecutionTimeout != 0)
+            {
+                builder.Append("--execution-timeout");
+                builder.AppendQuoted(settings.ExecutionTimeout.ToString());
```
To avoid any localization issues this should probably be `builder.AppendQuoted(settings.ExecutionTimeout.ToString(CultureInfo.InvariantCulture));`

devlead commented:
@gep13 looks great just left a minor comment.

gep13 commented:
@devlead @patriksvensson ok, I have just added the ability to do a ChocolateyInstall alias :smile_cat: 

Seems to be working for me running it locally.

patriksvensson commented:
@gep13 It looks good to me :smile:

devlead commented:
@patriksvensson  agree! @gep13 would you like implement rest methods in this or in later PR?

gep13 commented:
Thank you, both of you! :smile: 

@devlead I am lost, what rest methods?

devlead commented:
@gep13 I meant commands. :)

gep13 commented:
@devlead  Ah, I gotcha.  I thought you were talking about REST methods, and I was like WTF! :smile_cat: 

I updated the first post in this PR to give an indication of what I was hoping to achieve in this PR.  If you want I can break these up into separate PR's.

devlead commented:
@gep13 oh no worries ;) Either way works, it's a new feature so don't see any issue with doing it in this PR. Just us know when you want something reviewed.

devlead commented:
@gep13 this looks good :+1:  , only thing I'm missing after push & update is usage sample documentation. Really helpful when testing new aliases, se [NuGetAliases.cs#L31](https://github.com/cake-build/cake/blob/develop/src/Cake.Common/Tools/NuGet/NuGetAliases.cs#L31) for reference,

gep13 commented:
@devlead thanks for taking a look!

Yip, that documentation is going to be the last thing that I do.  I have already changed the parameters for the aliases twice already during some refactoring, so didn't want the hassle of updating the documentation as I was going. :smiley_cat:

RichiCoder1 commented on src/Cake.Common/Tools/Chocolatey/ApiKey/ChocolateyApiKeySetter.cs:
```diff
@@ -0,0 +1,121 @@
+
+            builder.Append("apikey");
+
+            builder.Append("-s");
```
I really need to get my append switch PR out :P

gep13 replied:
@RichiCoder1 what does that do?  I am a Cake noob remember :smile_cat:

RichiCoder1 replied:
Nothing currently, you're good.
There's just a lot of code like:
```csharp
builder.Append("--switch");
builder.Append(switchValue);
```
Soo I wanted to introduce something like
```csharp
builder.AppendSwitch("--switch", switchValue);
```

RichiCoder1 commented:
Seems to be an awful lot of redundancy but I'm not sure if there's any good way to address that. Otherwise looks very solid.

gep13 commented:
Are you talking about the common command line parameters? If so, I was thinking the same. Was hoping to address at least some of it, once all the commands are in, and I can see where we are.

Sent from my Windows Phone
________________________________
From: Richard Simpson<mailto:notifications@github.com>
Sent: ‎09/‎10/‎2015 18:27
To: cake-build/cake<mailto:cake@noreply.github.com>
Cc: Gary Ewan Park<mailto:gep13@gep13.co.uk>
Subject: Re: [cake] [WIP] Chocolatey Addin for Cake (#434)

Seems to be an awful lot of redundancy but I'm not sure if there's any good way to address that. Otherwise looks very solid.

---
Reply to this email directly or view it on GitHub:
https://github.com/cake-build/cake/pull/434#issuecomment-146939225

gep13 commented:
@patriksvensson @devlead Ok, I "think" short of some refactoring into common base methods, which I am not sure I want to do just now, until this has been road tested, this PR is good to go.

Can you have a look at let me know what you think?

Cheers!

devlead commented on src/Cake.Common/Tools/Chocolatey/ChocolateyAliases.cs:
```diff
@@ -0,0 +1,830 @@
+        ///                                     AllowUnoffical          = false
+        ///                                 };
+        ///
+        ///     ChocolateyPack(chocolateyPackSettings);
```
Question is standard nuspec XSD valid in this case or is there an error? I ran this code and got:
```
Chocolatey v0.9.9.1
The element 'metadata' in namespace 'http://schemas.microsoft.com/packaging/2010/07/nuspec.xsd' has inval
id child element 'packageSourceUrl' in namespace 'http://schemas.microsoft.com/packaging/2010/07/nuspec.x
sd'. List of possible elements expected: 'licenseUrl, references, tags, dependencies, language, developme
ntDependency, frameworkAssemblies, iconUrl, releaseNotes, requireLicenseAcceptance, copyright' in namespa
ce 'http://schemas.microsoft.com/packaging/2010/07/nuspec.xsd'.
Error: Chocolatey: Process returned an error.
```

gep13 replied:
Sorry, you ran this code where?  And with what version of Chocolatey installed?  Chocolatey has now added some additional parameters to the nuspec file which will no longer work if trying to use `nuget pack`, only the `choco pack` will work.  But again, only with newer versions of Chocolatey.  I was using the latest 0.9.9.11 for testing the Addin, and I guess this dependency should be recorded somewhere.

devlead replied:
Ah was on `Chocolatey v0.9.9.1` upgrading to `Chocolatey v0.9.9.11` and now it works.

gep13 replied:
Cool :-)

Where should this dependency be noted?

Sent from my Windows Phone
________________________________
From: Mattias Karlsson<mailto:notifications@github.com>
Sent: ‎11/‎10/‎2015 18:15
To: cake-build/cake<mailto:cake@noreply.github.com>
Cc: Gary Ewan Park<mailto:gep13@gep13.co.uk>
Subject: Re: [cake] [WIP] Chocolatey Addin for Cake (#434)

> +        ///                                     IconUrl                 = new Uri("http://cdn.rawgit.com/SomeUser/TestChocolatey/master/icons/testchocolatey.png"),
> +        ///                                     ReleaseNotes            = new [] {"Bug fixes", "Issue fixes", "Typos"},
> +        ///                                     Files                   = new [] {
> +        ///                                                                          new ChocolateyNuSpecContent {Source = "bin/TestChocolatey.dll", Target = "bin"},
> +        ///                                                                       },
> +        ///                                     Debug                   = false,
> +        ///                                     Verbose                 = false,
> +        ///                                     Force                   = false,
> +        ///                                     Noop                    = false,
> +        ///                                     LimitOutput             = false,
> +        ///                                     ExecutionTimeout        = 13,
> +        ///                                     CacheLocation           = @"C:\temp",
> +        ///                                     AllowUnoffical          = false
> +        ///                                 };
> +        ///
> +        ///     ChocolateyPack(chocolateyPackSettings);

Ah was on `Chocolatey v0.9.9.1` upgrading to `Chocolatey v0.9.9.11` and now it works.

---
Reply to this email directly or view it on GitHub:
https://github.com/cake-build/cake/pull/434/files#r41712943

devlead replied:
@gep13 perhaps add min supported version on `ChocolateyPackSettings` properties.

gep13 replied:
@devlead not sure I fully understand what you mean.  Can you provide an example?

devlead replied:
Nothing fancy, perhaps just add doc remarks on the settings class properties that require a minimal Chocolatey version?
I.e. this was one of the one of the previous version complains about:
[ChocolateyPackSettings.cs#L64]
(https://github.com/gep13/cake/blob/ChocolateyAddin/src/Cake.Common/Tools/Chocolatey/Pack/ChocolateyPackSettings.cs#L64)

gep13 replied:
Gotcha. I have added that in now :smile:

devlead replied:
@gep13 epic! :+1:

devlead commented:
@gep13 think this PR looks very solid, you've done great work here :+1: If you don't have anything major I think it's time to remove` [WIP]` and rebase against latest develop.

gep13 commented:
Done and done :+1:

devlead commented:
@gep13 this looks good could you rebase into 1 commit?

gep13 commented:
I certainly can if that is your preference.  I just thought it was a lot of code to pull in, in 1 commit.

patriksvensson commented:
@devlead I agree with @gep13 here. The PR:s are specialized enough. I would however want them reworded if it's possible. Especially the last one which should be something like `Added documentation for Chocolatey.` or similar.

gep13 commented:
@patriksvensson happy to re-word in anyway you want.

What would you want the individual commits re-worded as?  Currently they are in the format of:

`(GH-325) Adding Pack Command with Tests`

What should be different here?

patriksvensson commented:
@devlead @gep13 And the other PR:s should include `Chocolatey` I think.

Something like: `Adding Chocolatey ApiKey command with tests.`

devlead commented:
@gep13 I could be fine with 10 commits, but in that case as @patriksvensson rewording is needed so git log is clear.

gep13 commented:
Sure, I can do that.

Just one question, though... Don't you like/want the related issue number in the commit message?  I personally find that really useful for jumping to the issue when looking at the commits.

patriksvensson commented:
@gep13 I'm OK with you leaving them there.
I prefer the format below, but I think your approach is good as well.

```
subject

description
description
description

Closes #123
```

devlead commented:
I've used the `closes #325` syntax in commit message "line2"

gep13 commented:
Yeah, I have used that syntax as well, but I prefer to take ownership over when the issue actually gets closed, typically by moving it between lanes in a kanban board of some kind.

I have rebased the commit messages, let me know what you think.

patriksvensson commented:
@gep13 Fair enough :smile: 
Everything looks good to me what I've seen so this should be OK to merge. :+1:

devlead commented:
@gep13 epic pr, thanks :+1:

patriksvensson commented:
@gep13 Agree. Super epic :smile:

gep13 commented:
Aw shucks! Thank you guys for Cake :-)
